_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+
+4. `@ContainedIn` just like `@OneToMany`-`@ManyToOne` annotation makes relationship bidirectional. When the values of this
+entity will be updated its values in the root entity i.e., in the `idx_post`(Post) index will also be updated.
+
+## Performing Queries
+Before we perform any queries, we first need to load data into the Elasticsearch.
+
+Use the following snippet to do the same.
+
+ ```java
+@Service
+@RequiredArgsConstructor
+@Slf4j
+class IndexingService {
+
+  private final EntityManager em;
+
+  @Transactional
+  public void initiateIndexing() throws InterruptedException {
+      log.info("Initiating indexing...");
+      FullTextEntityManager fullTextEntityManager = 
+                            Search.getFullTextEntityManager(em);
+      fullTextEntityManager.createIndexer().startAndWait();
+      log.info("All entities indexed");
+  }
+}
+```
+
+You can call the `initiateIndexing()` function either at the application startup or create an API in the REST controller to call it.
+`createIndexer` function also takes in Class references as input. This gives you more choice over the indexing. This is going to be a one-time thing, after this Hibernate search will keep entities in both sources in sync. Unless of course for some reason your database goes out of sync with Elasticsearch in which case indexing API might come in handy again. 
+
+With Elasticsearch integration you have two choices for writing queries:
+1. Hibernate search query DSL - It's a nice way to write Lucene queries. If you are familiar with Specification and 
+Criteria query API you will find it easy to get your head around it.
+2. Elasticsearch query - Hibernate search supports both Elasticsearch native queries and JSON queries.
+
+In this tutorial, we are only going to look at Hibernate Search query DSL. 
+
+Now let's say we want to write a query to fetch all records from `idx_post`  where either `body` or `hashtags` contain 
+word 'food'
+
+```java
+@Component
+@Slf4j
+@RequiredArgsConstructor
+public class SearchService {
+
+  private final EntityManager entityManager;
+  //.........
+  public List<Post> getPostBasedOnWord(String word){
+    FullTextEntityManager fullTextEntityManager = 
+              Search.getFullTextEntityManager(entityManager);
+
+    QueryBuilder qb = fullTextEntityManager
+              .getSearchFactory()
+              .buildQueryBuilder()
+              .forEntity(Post.class)
+              .get();
+
+    Query foodQuery = qb.keyword()
+              .onFields("body","hashTags")
+              .matching(word)
+              .createQuery();
+
+    FullTextQuery fullTextQuery = fullTextEntityManager
+                          .createFullTextQuery(foodQuery, Post.class);
+    return (List<Post>) fullTextQuery.getResultList();
+  }
+  //.........
+  
+}
+```
+
+Let me explain the above code line by line:
```

Comment:
This sentence comes across as arrogant (I know it's not meant that way). I replaced it.